src/core/components/pin.js

```diff
@@ -0,0 +1,412 @@
+      })
+    }),
+
+    // encodes and writes pin key sets to the datastore
+    // each key set will be stored as a DAG node, and a root node will link to both
+    flush: promisify((callback) => {
+      const handle = {
+        put: (k, v, cb) => {
+          handle[k] = v
+          cb()
+        }
+      }
+
+      waterfall([
+        // create link to direct keys node
+        (cb) => pin.set.storeSet(pin.directKeys(), cb),
+        (dRoot, cb) => DAGLink.create(pin.types.direct, dRoot.size, dRoot.multihash, cb),
+        (dLink, cb) => handle.put('dLink', dLink, cb),
+
+        // create link to recursive keys node
+        (cb) => pin.set.storeSet(pin.recursiveKeys(), cb),
+        (rRoot, cb) => DAGLink.create(pin.types.recursive, rRoot.size, rRoot.multihash, cb),
+        (rLink, cb) => handle.put('rLink', rLink, cb),
+
+        // the pin-set nodes link to an empty node, so make sure it's added to dag
+        (cb) => DAGNode.create(Buffer.alloc(0), cb),
+        (empty, cb) => dag.put(empty, {cid: new CID(empty.multihash)}, cb),
+
+        // create root node with links to direct and recursive nodes
+        (cid, cb) => DAGNode.create(Buffer.alloc(0), [handle.dLink, handle.rLink], cb),
+        (root, cb) => handle.put('root', root, cb),
+
+        // add the root node to dag
+        (cb) => dag.put(handle.root, {cid: new CID(handle.root.multihash)}, cb),
+
+        // save serialized root to datastore under a consistent key
+        (_, cb) => repo.closed ? repo.datastore.open(cb) : cb(null, null), // hack for CLI tests
+        (_, cb) => repo.datastore.put(pinDataStoreKey, handle.root.multihash, cb)
+      ], (err, result) => {
+        if (err) { return callback(err) }
+        self.log(`Flushed pins with root: ${handle.root}.`)
+        return callback(null, handle.root)
+      })
+    }),
+
+    load: promisify(callback => {
+      const handle = {
+        put: (k, v, cb) => {
+          handle[k] = v
+          cb()
+        }
+      }
+
+      waterfall([
+        (cb) => repo.closed ? repo.datastore.open(cb) : cb(null, null), // hack for CLI tests
+        (_, cb) => repo.datastore.has(pinDataStoreKey, cb),
+        (has, cb) => has ? cb() : cb(new Error('No pins to load')),
+        (cb) => repo.datastore.get(pinDataStoreKey, cb),
+        (mh, cb) => dag.get(new CID(mh), cb),
+        (root, cb) => handle.put('root', root.value, cb),
+        (cb) => pin.set.loadSet(handle.root, pin.types.recursive, cb),
+        (rKeys, cb) => handle.put('rKeys', rKeys, cb),
+        (cb) => pin.set.loadSet(handle.root, pin.types.direct, cb)
+      ], (err, dKeys) => {
+        if (err && err.message !== 'No pins to load') {
+          return callback(err)
+        }
+        if (dKeys) {
+          directPins = new Set(dKeys.map(toB58String))
+          recursivePins = new Set(handle.rKeys.map(toB58String))
+        }
+        self.log('Loaded pins from the datastore')
+        return callback()
+      })
+    })
+  }
+  return pin
```

Comment:
There's a lot of properties on this object that aren't part of the spec'd public API. If they're not for public consumption we need to indicate that by prefixing them with `_` or better still pull them out into modules/functions that can only be used internally.

Reply:
I've pulled some properties out entirely and `_`-spaced a few others. We now have:

`types`
`add`
`rm`
`ls`
`_load`
`_isPinnedWithType`

`_load` is used by ipfs core to initialize the pinset. `_isPinnedWithType` is not strictly necessary but seems to be a strong feature for developers working with the core modules.

Reply:
👍 

The pin types are constants that don't _need_ to be attached to an instance, I'd export them separately so that they can be used like:
`import { PinTypes } from 'ipfs'` or `const { PinTypes } = require('ipfs')`.